lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -1679,22 +1643,126 @@ void SessionState::get_charging_updates(
   }
 }
 
-void SessionState::fill_service_action(
+optional<CreditUsageUpdate> SessionState::get_update_for_continue_service(
+    const CreditKey& key, std::unique_ptr<ChargingGrant>& grant,
+    SessionStateUpdateCriteria& session_uc) {
+  CreditUsage::UpdateType update_type;
+  if (!grant->get_update_type(&update_type)) {
+    return {};  // no update
+  }
+  if (curr_state_ == SESSION_RELEASED) {
+    MLOG(MDEBUG) << "Session " << session_id_
+                 << " is in Released state. Not sending update to the core"
+                    "for rating group "
+                 << key;
+    return {};  // no update
+  }
+  if (grant->suspended && update_type == CreditUsage::QUOTA_EXHAUSTED) {
+    MLOG(MDEBUG) << "Credit " << key << " for " << session_id_
+                 << " is suspended. Not sending update to the core";
+    return {};  // no update
+  }
+
+  // Create Update struct
+  MLOG(MDEBUG) << "Subscriber " << imsi_ << " rating group " << key
+               << " updating due to type "
+               << credit_update_type_to_str(update_type)
+               << " with request number " << request_number_;
+
+  auto credit_uc = get_credit_uc(key, session_uc);
+  if (update_type == CreditUsage::REAUTH_REQUIRED) {
+    grant->set_reauth_state(REAUTH_PROCESSING, *credit_uc);
+  }
+  CreditUsage usage = grant->get_credit_usage(update_type, *credit_uc, false);
+  key.set_credit_usage(&usage);
+
+  auto request = make_credit_usage_update_req(usage);
+  request_number_++;
+  session_uc.request_number_increment++;
+  return request;
+}
+
+void SessionState::fill_service_action_for_activate(
+    std::unique_ptr<ServiceAction>& action_p, const CreditKey& key) {
```

Comment:
not sure if there is a convention here, but since `action_p` is the object modified/output argument, shouldn't we put it as the last argument?

Reply:
yeah actually, maybe it might be good to just attach these functions to the Action object.... Will fix in one of the stacked changes to avoid a nasty rebase.